### PR TITLE
NetCore: Fixes registration issue (continued)

### PR DESCRIPTION
**Pull request by rustybox**
Add additional "mutliple unique" registrations and rename extension for clarity.

#9323

### CODE REVIEW

**Comment by rustybox** [comment]
@bergmania @AndyButland good 2 merge?

**Comment by bergmania** [comment]
Thanks @rustybox

**Comment by rustybox** [comment]
I do wonder if this is worse than original implementation of RegisterUnique

```csharp
composition.RegisterUnique<TImplementing>();
composition.RegisterUnique<TService1>(factory => factory.GetInstance<TImplementing>());
composition.RegisterUnique<TService2>(factory => factory.GetInstance<TImplementing>());
```

Imagine someone comes along now and re-registers TService1 but with a concrete that doesn't implement TService2, everything will seem fine but the cast will be invalid when resolving TService2.

On the other hand, I imagine this extension is rarely used except when both interfaces are tightly coupled together as hinted by with Andy's issue where calls to one was expected to affect the other. I think this suggests that the interface boundaries are weird, if a method call on TService1 should affect state on TService2 perhaps the two operations should be as part of the same interface.

In short I don't really know what's right and wrong here but there is a potential for confusion later if anyone ever replaces TService1 without replacing TService2

**Comment by bergmania** [comment]
Good points @rustybox .

I don't like the idea about merging the interfaces, however, I think i should be pretty easy to avoid this, if we just split the implementation in two.

E.g.
I think it is simple to split `AspNetCoreSessionManager` into a new `AspNetCoreSessionManager` wich does not implement `ISessionIdResolver` and a `AspNetCoreSessionIdResolver` that implements `ISessionIdResolver`.

If required we can inject `ISessionIdResolver` into `AspNetCoreSessionManager`. But I'm not sure this is even required. The main reason this is just one implementation is that it basically needs the same dependencies, IIRC.

The same approach can be done with the other 2 types where we use `AddMultipleUnique`

**Comment by rustybox** [comment]
@bergmania In the short term do you think we need to revert to how it was previously where concrete is registered or do you think it is relatively safe as is?

**Comment by bergmania** [comment]
I think it is relatively safe as is.
